File: go/program/ops/opVariable.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
	"github.com/2manyvcos/jpl/go/library"
)

type opVariable struct{}

// { name: string }
func (opVariable) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	value, ok := scope.Vars()[params.Name]
	if !ok {
		return nil, library.ThrowAny(library.NewReferenceError("%s is not defined", params.Name))
	}
	return next.Pipe(value, scope)
}

// { name: string }
func (opVariable) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (definition.JPLInstructionParams, jpl.JPLError) {
	return definition.JPLInstructionParams{
		Name: params.Name,
	}, nil
}

File: js/lib/program/ops/opVariable.ts
import type { JPLInstructionParams } from '@/library/definition';
import type JPLRuntime from '@/runtime';
import { JPLReferenceError, type JPLRuntimeScope } from '../../library';
import type { UserJPLInstructionParams } from '../params';

export default {
  /** { name: string } */
  op(
    _runtime: JPLRuntime,
    _input: unknown,
    params: JPLInstructionParams,
    scope: JPLRuntimeScope,
    next: (
      output: unknown,
      scope: JPLRuntimeScope,
    ) => Promise<unknown[]> | unknown[],
  ): Promise<unknown[]> | unknown[] {
    if (!Object.hasOwn(scope.vars, params.name ?? ''))
      throw new JPLReferenceError('%s is not defined', params.name ?? '');
    return next(scope.vars[params.name ?? ''], scope);
  },

  /** { name: string } */
  map(
    runtime: JPLRuntime,
    params: UserJPLInstructionParams,
  ): JPLInstructionParams {
    return {
      name: runtime.assertType(params.name, 'string'),
    };
  },
};
